File: mobile/src/main/java/org/openhab/habdroid/util/SyncHttpClient.java
/*
 * Copyright (c) 2010-2016, openHAB.org and others.
 *
 *   All rights reserved. This program and the accompanying materials
 *   are made available under the terms of the Eclipse Public License v1.0
 *   which accompanies this distribution, and is available at
 *   http://www.eclipse.org/legal/epl-v10.html
 */

package org.openhab.habdroid.util;

import android.content.Context;
import android.content.SharedPreferences;

import java.io.IOException;
import java.util.Map;

import okhttp3.Call;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.ResponseBody;

public class SyncHttpClient extends HttpClient {
    public static class HttpResult {
        public final Request request;
        public final ResponseBody response;
        public final Throwable error;
        public final int statusCode;

        private HttpResult(Call call) {
            ResponseBody result = null;
            Throwable error = null;
            int code = 500;

            try {
                Response response = call.execute();
                code = response.code();
                result = response.body();
                if (!response.isSuccessful()) {
                    error = new IOException(response.code() + ": " + response.message());
                }
            } catch (IOException e) {
                error = e;
            }

            this.statusCode = code;
            this.request = call.request();
            this.response = result;
            this.error = error;
        }

        public boolean isSuccessful() {
            return error == null;
        }
        public void close() {
            if (response != null) {
                response.close();
            }
        }

        public HttpTextResult asText() {
            return new HttpTextResult(this);
        }
        public HttpStatusResult asStatus() {
            return new HttpStatusResult(this);
        }
    }

    public static class HttpStatusResult {
        public final Throwable error;
        public final int statusCode;

        HttpStatusResult(HttpResult result) {
            this.error = result.error;
            this.statusCode = result.statusCode;
            result.close();
        }

        public boolean isSuccessful() {
            return error == null;
        }
    }

    public static class HttpTextResult {
        public final Request request;
        public final String response;
        public final Throwable error;
        public final int statusCode;

        HttpTextResult(HttpResult result) {
            this.request = result.request;
            this.statusCode = result.statusCode;
            if (result.response == null) {
                this.response = null;
                this.error = result.error;
            } else {
                String response = null;
                Throwable error = result.error;
                try {
                    response = result.response.string();
                } catch (IOException e) {
                    error = e;
                }
                this.response = response;
                this.error = error;
            }
            result.close();
        }

        public boolean isSuccessful() {
            return error == null;
        }
    }

    public SyncHttpClient(Context context, String baseUrl) {
        super(context, baseUrl);
    }

    public SyncHttpClient(Context context, SharedPreferences prefs, String baseUrl) {
        super(context, prefs, baseUrl);
    }

    public HttpResult get(String url) {
        return get(url, null);
    }

    public HttpResult get(String url, Map<String, String> headers) {
        return method(url, "GET", headers, null, null);
    }

    public HttpResult post(String url, String requestBody, String mediaType) {
        return method(url, "POST", null, requestBody, mediaType);
    }

    protected HttpResult method(String url, String method, Map<String, String> headers,
            String requestBody, String mediaType) {
        return new HttpResult(
                prepareCall(url, method, headers, requestBody, mediaType, CachingMode.AVOID_CACHE));
    }
}

File: mobile/src/main/java/org/openhab/habdroid/util/SyncHttpClient.kt
/*
 * Copyright (c) 2010-2016, openHAB.org and others.
 *
 *   All rights reserved. This program and the accompanying materials
 *   are made available under the terms of the Eclipse Public License v1.0
 *   which accompanies this distribution, and is available at
 *   http://www.eclipse.org/legal/epl-v10.html
 */

package org.openhab.habdroid.util

import okhttp3.Call
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.ResponseBody
import java.io.IOException

class SyncHttpClient(client: OkHttpClient, baseUrl: String?, username: String?, password: String?) :
        HttpClient(client, baseUrl, username, password) {
    class HttpResult constructor(call: Call) {
        val request: Request
        val response: ResponseBody?
        val error: Throwable?
        val statusCode: Int

        val isSuccessful: Boolean
            get() = error == null

        init {
            var result: ResponseBody? = null
            var error: Throwable? = null
            var code = 500

            try {
                val response = call.execute()
                code = response.code()
                result = response.body()
                if (!response.isSuccessful) {
                    error = IOException(response.code().toString() + ": " + response.message())
                }
            } catch (e: IOException) {
                error = e
            }

            this.statusCode = code
            this.request = call.request()
            this.response = result
            this.error = error
        }

        fun close() {
            response?.close()
        }

        fun asText(): HttpTextResult {
            return HttpTextResult(this)
        }

        fun asStatus(): HttpStatusResult {
            return HttpStatusResult(this)
        }
    }

    class HttpStatusResult internal constructor(result: HttpResult) {
        val error: Throwable? = result.error
        val statusCode: Int = result.statusCode

        val isSuccessful: Boolean
            get() = error == null

        init {
            result.close()
        }
    }

    class HttpTextResult internal constructor(result: HttpResult) {
        val request = result.request
        val statusCode = result.statusCode
        val response: String?
        val error: Throwable?

        val isSuccessful: Boolean
            get() = error == null

        init {
            if (result.response == null) {
                this.response = null
                this.error = result.error
            } else {
                var response: String? = null
                var error = result.error
                try {
                    response = result.response.string()
                } catch (e: IOException) {
                    error = e
                }

                this.response = response
                this.error = error
            }
            result.close()
        }
    }

    operator fun get(url: String): HttpResult {
        return get(url, -1)
    }

    operator fun get(url: String, timeoutMillis: Long): HttpResult {
        return method(url, "GET", null, null, null, timeoutMillis)
    }

    operator fun get(url: String, headers: Map<String, String>? = null): HttpResult {
        return method(url, "GET", headers, null, null, -1)
    }

    fun post(url: String, requestBody: String,
             mediaType: String, headers: Map<String, String>? = null): HttpResult {
        return method(url, "POST", headers, requestBody, mediaType, -1)
    }

    protected fun method(url: String, method: String, headers: Map<String, String>?,
                         requestBody: String?, mediaType: String?, timeoutMillis: Long): HttpResult {
        val call = prepareCall(url, method, headers, requestBody,
                mediaType, timeoutMillis, CachingMode.AVOID_CACHE)
        return HttpResult(call)
    }
}
